### PR TITLE
Refactor VS Code auth

### PR DESCRIPTION
**Pull request by alexweininger**
Refactor VS Code auth portion of the vscode.dev work to hide the auth implementation details from the rest of the code. Now both the VS Code auth (used in vscode.dev) and the Azure Account API auth are shoehorned into a common interface.

This makes it easier to adapt for testing purposes and is similar to what I've done in https://github.com/microsoft/vscode-azureresourcegroups/pull/563

### CODE REVIEW

File: src/services/DesktopSubscriptionProvider.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+export function createAzureAccountSubscriptionProviderFactory(): () => Promise<AzureSubscriptionProvider> {
+    return async () => {
+        const api = await getAzureAccountExtensionApi();
+        return new AzureAccountSubscriptionProvider(api);
+    }
+}
+
+export class AzureAccountSubscriptionProvider implements AzureSubscriptionProvider {
```
Comment (bwateratmsft) [inline src/services/DesktopSubscriptionProvider.ts:38]: Does this need to be exported still?

File: src/extension.ts [hunk 2]
```diff
@@ -64,7 +65,7 @@ export async function activate(context: vscode.ExtensionContext, perfStats: { lo
         activateContext.telemetry.measurements.mainFileLoad = (perfStats.loadEndTime - perfStats.loadStartTime) / 1000;
 
         setupEvents(context);
-        ext.subscriptionProvider = new VSCodeAzureSubscriptionProvider(context.globalState);
+        ext.subscriptionProviderFactory = ext.isWeb ? createWebSubscriptionProviderFactory(context) : createAzureAccountSubscriptionProviderFactory();
```
Comment (bwateratmsft) [inline src/extension.ts:68]:
```suggestion
        ext.subscriptionProviderFactory ??= ext.isWeb ? createWebSubscriptionProviderFactory(context) : createAzureAccountSubscriptionProviderFactory();
```

Might be useful for unit tests? I dunno.

**Review by bwateratmsft — APPROVED** [review]

File: src/extension.ts (hunk 2 above)
Comment (alexweininger) [reply src/extension.ts:68]: For unit tests, what's worked best for me is adding a `ext.overrideBlahFactory`.

File: src/services/DesktopSubscriptionProvider.ts (hunk 1 above)
Comment (alexweininger) [reply src/services/DesktopSubscriptionProvider.ts]: No, it doesn't

**Review by bwateratmsft — APPROVED** [review]

**Comment by nturinski** [comment]
I think you need to revert some of the `package.json` changes.

This error message: `12 verbose stack Invalid: lock file's @microsoft/vscode-azext-utils@0.5.1 does not satisfy @microsoft/vscode-azext-utils@0.4.3-alpha` is preventing it from building.

You can use 0.5.2-alpha.0. Should have all the changes of 0.5.1 + the dev stuff required.

**Comment by alexweininger** [comment]
What's going on with this build? 😢

**Comment by nturinski** [comment]
I've never seen this error before:
```
C:\Windows\system32\cmd.exe /D /S /C "C:\hostedtoolcache\windows\node\14.21.3\x64\npm.cmd ci"
npm ERR! assert not accessible from @microsoft/vscode-azext-dev
```

**Comment by nturinski** [comment]
So annoying... I couldn't test it because of the dependency on the Azure Accounts extension in the `package.json`. I can manually change it, but should definitely do something about it on the scripting level.

Anyway, get these two errors when testing it out:

```
Cannot read properties of undefined (reading 'length')
this.subscriptionResultsTask is not a function
```

**Review by nturinski — APPROVED** [review]

**Pull request merged**